libraries/Adapters/Microsoft.Bot.Builder.Adapters.Facebook/FacebookClientWrapper.cs

```diff
@@ -61,6 +64,78 @@ public async Task<HttpResponseMessage> CallAPIAsync(string path, FacebookMessage
+        /// This is used by many internal functions to get access to the Facebook API, and is exposed as `bot.api` on any BotWorker instances passed into Botkit handler functions.
+        /// </summary>
+        /// <param name="activity">An incoming message activity.</param>
+        /// <returns>A Facebook API client.</returns>
+        public async Task<FacebookClientWrapper> GetAPIAsync(Activity activity)
+        {
+            if (!string.IsNullOrWhiteSpace(Options.AccessToken))
+            {
+                return new FacebookClientWrapper(new FacebookAdapterOptions(Options.VerifyToken, Options.AppSecret, Options.AccessToken));
+            }
+            else
+            {
+                if (!string.IsNullOrWhiteSpace(activity.Recipient?.Id))
+                {
+                    var pageId = activity.Recipient.Id;
+
+                    if ((activity.ChannelData as dynamic)?.message != null && (activity.ChannelData as dynamic)?.message.is_echo)
+                    {
+                        pageId = activity.From.Id;
+                    }
+
+                    var token = await Options.GetAccessTokenForPageAsync(pageId).ConfigureAwait(false);
+
+                    if (string.IsNullOrWhiteSpace(token))
+                    {
+                        // error: missing credentials
+                    }
+
+                    return new FacebookClientWrapper(new FacebookAdapterOptions(Options.VerifyToken, Options.AppSecret, token));
+                }
+                else
+                {
+                    throw new Exception($"Unable to create API based on activity:{activity}");
+                }
+            }
+        }
+
+        /// <summary>
+        /// Verifies the SHA1 signature of the raw request payload before bodyParser parses it will abort parsing if signature is invalid, and pass a generic error to response.
+        /// </summary>
+        /// <param name="request">An Http request object.</param>
+        /// <param name="response">An Http response object.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public async Task<bool> VerifySignatureAsync(HttpRequest request, HttpResponse response, CancellationToken cancellationToken)
+        {
+            var expected = request.Headers["x-hub-signature"];
+
+            string calculated = null;
+            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(Options.AppSecret)))
+            {
+                using (var bodyStream = new StreamReader(request.Body))
+                {
+                    calculated = $"sha1={hmac.ComputeHash(Encoding.UTF8.GetBytes(bodyStream.ReadToEnd()))}";
+                }
+            }
+
+            if (expected != calculated)
+            {
+                response.StatusCode = 401;
```

Comment:
See if you can replace it by an HttpStatusCode. This was feedback from previous adapters. Something like
`response.StatusCode = Convert.ToInt32(HttpStatusCode.Unauthorized, System.Globalization.CultureInfo.InvariantCulture);`